lambdas/process-candidate-identity/src/main/java/uk/gov/di/ipv/core/processcandidateidentity/service/CheckCoiService.java

```diff
@@ -0,0 +1,204 @@
+            IpvSessionItem ipvSessionItem,
+            ClientOAuthSessionItem clientOAuthSession,
+            CoiCheckType checkType,
+            String deviceInformation,
+            String ipAddress,
+            List<VerifiableCredential> sessionVcs)
+            throws HttpResponseExceptionWithErrorBody, CredentialParseException,
+                    EvcsServiceException {
+
+        var ipvSessionId = ipvSessionItem.getIpvSessionId();
+        var userId = clientOAuthSession.getUserId();
+        var govukSigninJourneyId = clientOAuthSession.getGovukSigninJourneyId();
+        LogHelper.attachGovukSigninJourneyIdToLogs(govukSigninJourneyId);
+
+        if (TRUE.equals(clientOAuthSession.getReproveIdentity())) {
+            LOGGER.info(
+                    LogHelper.buildLogMessage(
+                            "Reprove identity flag set - checking full name and DOB"));
+            checkType = ACCOUNT_INTERVENTION;
+        }
+
+        var auditEventUser =
+                new AuditEventUser(userId, ipvSessionId, govukSigninJourneyId, ipAddress);
+        sendAuditEvent(
+                AuditEventTypes.IPV_CONTINUITY_OF_IDENTITY_CHECK_START,
+                checkType,
+                null,
+                auditEventUser,
+                null,
+                null,
+                deviceInformation);
+
+        var oldVcs =
+                evcsService.getVerifiableCredentials(
+                        userId, clientOAuthSession.getEvcsAccessToken(), CURRENT);
+        var combinedCredentials = Stream.concat(oldVcs.stream(), sessionVcs.stream()).toList();
+        var successfulCheck =
+                switch (checkType) {
+                    case STANDARD -> userIdentityService.areNamesAndDobCorrelated(
+                            combinedCredentials);
+                    case REVERIFICATION -> userIdentityService
+                            .areNamesAndDobCorrelatedForReverification(combinedCredentials);
+                    case ACCOUNT_INTERVENTION -> userIdentityService.areVcsCorrelated(
+                            combinedCredentials);
+                };
+
+        var scopeClaims = clientOAuthSession.getScopeClaims();
+        var isReverification = scopeClaims.contains(ScopeConstants.REVERIFICATION);
+
+        sendAuditEvent(
+                AuditEventTypes.IPV_CONTINUITY_OF_IDENTITY_CHECK_END,
+                checkType,
+                successfulCheck,
+                auditEventUser,
+                oldVcs,
+                sessionVcs,
+                deviceInformation);
+
+        if (isReverification) {
+            setIpvSessionReverificationStatus(
+                    ipvSessionItem,
+                    successfulCheck ? ReverificationStatus.SUCCESS : ReverificationStatus.FAILED);
+            if (!successfulCheck) {
+                ipvSessionItem.setFailureCode(ReverificationFailureCode.IDENTITY_DID_NOT_MATCH);
+                ipvSessionService.updateIpvSession(ipvSessionItem);
```

Comment:
I should have spotted this before (or maybe it was a slightly dodgy merge), but we're doubling up the DynamoDB calls here...

It could probably just be:
```
if (isReverification) {
    ipvSession.setReverificationStatus(successfulCheck
            ? ReverificationStatus.SUCCESS
            : ReverificationStatus.FAILED);
    if (!successfulCheck) {
        ipvSession.setFailureCode(ReverificationFailureCode.IDENTITY_DID_NOT_MATCH);
    }
    ipvSessionService.updateIpvSession(ipvSession);
}
```

Reply:
(The `updateIpvSession` is this bit that actually updates it in DynamoDB)

Reply:
ahh thanks for the spot! I've removed the `ipvSessionService.updateIpvSession(ipvSession);` from here though bc there's one in `finally` block in the [handler](https://github.com/govuk-one-login/ipv-core-back/blob/d82922f6a8dacae0307ebe3906f21c54977babda/lambdas/process-candidate-identity/src/main/java/uk/gov/di/ipv/core/processcandidateidentity/ProcessCandidateIdentityHandler.java#L227)

Reply:
actually scratch that, maybe we don't want to do a save for every call to this lambda if no updates were made (though it's just extra safety incase we miss an update)? I think only store identity and coi are updating the ipv session item

Reply:
Yeah - it's not the end of the world to update more than strictly necessary (and we definitely don't want to risk dropping an update), but this case seems particularly easy to optimise!

This is the sort of thing where frameworks can help - they'll track when the record is 'dirty' (i.e. modified) and save conditionally, but I don't think we really need anything so complicated.